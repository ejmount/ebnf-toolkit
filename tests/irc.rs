use ebnf_toolkit::Grammar;
use insta::assert_compact_debug_snapshot;

static SRC: &str = r#"
message       ::= ['@' tags SPACE] [':' source SPACE ] command [parameters] crlf;
tags          ::= tag [';' tag]*;
tag           ::= key ['=' escaped_value];
key           ::= [ client_prefix ] [ vendor '/' ] key_name;
vendor        ::= #'[a-zA-Z0-9]+';
key_name      ::= #'[a-zA-Z0-9]+';
command       ::= #'[a-zA-Z0-9]+';
escaped_value ::= #'[a-zA-Z0-9]+';
client_prefix ::= '+';
source          ::=  servername | username;
username        ::=  ( nick [ '!' user ] [ '@' host ] );
nick            ::=  #'[^ \\0\r\n #][^\\0\r\n ]*';
user            ::=  #'[^\r\n ]';
servername      ::=  #'[a-zA-Z0-9]+';
host            ::=  #'[a-zA-Z0-9.]+';
parameters      ::=  ( SPACE middle )* [ SPACE ':' trailing ];
middle          ::=  nospcrlfcl ( ':' | nospcrlfcl )*;
trailing        ::=  ( ':' | ' ' | nospcrlfcl )*;
nospcrlfcl      ::=  #'[^ :\r\n]';
SPACE           ::= ' '+;
crlf            ::= '\r\n';
"#;

#[test]
fn irc_grammar() {
    let g = Grammar::new(SRC).unwrap_or_else(|e| panic!("{e}"));

    assert_eq!(g.get_dangling_reference(), None);

    let rules = g.rules;

    assert_compact_debug_snapshot!(rules.get("tags").unwrap());
    assert_compact_debug_snapshot!(rules.get("tag").unwrap());
    assert_compact_debug_snapshot!(rules.get("key").unwrap());
    assert_compact_debug_snapshot!(rules.get("vendor").unwrap());
    assert_compact_debug_snapshot!(rules.get("key_name").unwrap());
    assert_compact_debug_snapshot!(rules.get("command").unwrap());
    assert_compact_debug_snapshot!(rules.get("escaped_value").unwrap());
    assert_compact_debug_snapshot!(rules.get("client_prefix").unwrap());
    assert_compact_debug_snapshot!(rules.get("source").unwrap());
    assert_compact_debug_snapshot!(rules.get("username").unwrap());
    assert_compact_debug_snapshot!(rules.get("nick").unwrap());
    assert_compact_debug_snapshot!(rules.get("user").unwrap());
    assert_compact_debug_snapshot!(rules.get("servername").unwrap());
    assert_compact_debug_snapshot!(rules.get("host").unwrap());
    assert_compact_debug_snapshot!(rules.get("parameters").unwrap());
    assert_compact_debug_snapshot!(rules.get("middle").unwrap());
    assert_compact_debug_snapshot!(rules.get("trailing").unwrap());
    assert_compact_debug_snapshot!(rules.get("nospcrlfcl").unwrap());
    assert_compact_debug_snapshot!(rules.get("SPACE").unwrap());
    assert_compact_debug_snapshot!(rules.get("crlf").unwrap());
}
