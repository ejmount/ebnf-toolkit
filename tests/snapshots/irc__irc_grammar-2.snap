---
source: tests/irc.rs
expression: "rules.get(\"tag\").unwrap()"
---
Rule { name: "tag", body: [Nonterminal { span: Span { start: 135, end: 138, line_offset_start: (4, 18), line_offset_end: (4, 21) }, name: "key" }, Optional { span: Span { start: 139, end: 158, line_offset_start: (4, 22), line_offset_end: (4, 41) }, body: [Terminal { span: Span { start: 140, end: 143, line_offset_start: (4, 23), line_offset_end: (4, 26) }, str: "=" }, Nonterminal { span: Span { start: 144, end: 157, line_offset_start: (4, 27), line_offset_end: (4, 40) }, name: "escaped_value" }] }] }
